src/library_manager/lib_manager.c

```diff
@@ -68,114 +68,133 @@ static int lib_manager_load_data_from_storage(void __sparse_cache *vma, void *s_
+	const struct ext_library *const ext_lib = ext_lib_get();
+	const uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	const uintptr_t load_offset = POINTER_TO_UINT(ext_lib->desc[lib_id]);
+	void *src, *va_base;
+	size_t size;
+	uint32_t flags;
+	int ret, idx;
+
+	for (idx = 0; idx < ARRAY_SIZE(mod->segment); ++idx) {
+		if (!mod->segment[idx].flags.r.load)
+			continue;
+
+		flags = 0;
+
+		if (mod->segment[idx].flags.r.code)
+			flags = SYS_MM_MEM_PERM_EXEC;
+		else if (!mod->segment[idx].flags.r.readonly)
+			flags |= SYS_MM_MEM_PERM_RW;
+
+		src = UINT_TO_POINTER(mod->segment[idx].file_offset + load_offset);
+		va_base = UINT_TO_POINTER(mod->segment[idx].v_base_addr);
+		size = mod->segment[idx].flags.r.length * PAGE_SZ;
+		ret = lib_manager_load_data_from_storage(va_base, src, size, flags);
+		if (ret < 0)
+			goto err;
+	}
+
+	return 0;
+
+err:
+	for (--idx; idx >= 0; --idx) {
+		if (!mod->segment[idx].flags.r.load)
+			continue;
+
+		va_base = UINT_TO_POINTER(mod->segment[idx].v_base_addr);
+		size = mod->segment[idx].flags.r.length * PAGE_SZ;
+		sys_mm_drv_unmap_region(va_base, size);
+	}
+
+	return ret;
+}
+
+static int lib_manager_unload_module(const struct sof_man_module *const mod)
+{
+	void *va_base;
+	size_t size;
+	uint32_t idx;
 	int ret;
 
-	st_text_size = st_text_size * PAGE_SZ;
-	st_rodata_size = st_rodata_size * PAGE_SZ;
+	for (idx = 0; idx < ARRAY_SIZE(mod->segment); ++idx) {
+		if (!mod->segment[idx].flags.r.load)
+			continue;
 
-	/* Copy Code */
-	ret = lib_manager_load_data_from_storage(va_base_text, src_txt, st_text_size,
-						 SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
-	if (ret < 0)
-		goto err;
+		va_base = UINT_TO_POINTER(mod->segment[idx].v_base_addr);
+		size = mod->segment[idx].flags.r.length * PAGE_SZ;
+		ret = sys_mm_drv_unmap_region(va_base, size);
+		if (ret < 0)
+			return ret;
+	}
 
-	/* Copy RODATA */
-	ret = lib_manager_load_data_from_storage(va_base_rodata, src_rodata,
-						 st_rodata_size, SYS_MM_MEM_PERM_RW);
-	if (ret < 0)
-		goto err;
+	return 0;
+}
 
-	/* There are modules marked as lib_code. This is code shared between several modules inside
-	 * the library. Load all lib_code modules with first none lib_code module load.
-	 */
-	if (!mod->type.lib_code)
-		ext_lib->mods_exec_load_cnt++;
-
-	if (ext_lib->mods_exec_load_cnt == 1) {
-		struct sof_man_module *module_entry =
-		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
-		for (size_t idx = 0; idx < desc->header.num_module_entries;
-				++idx, ++module_entry) {
-			if (module_entry->type.lib_code) {
-				ret = lib_manager_load_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
-							      idx, mod, desc);
-				if (ret < 0)
-					goto err;
-			}
+/* There are modules marked as lib_code. This is code shared between several modules inside
+ * the library. Load all lib_code modules with first none lib_code module load.
+ */
+static int lib_manager_load_code_modules(const uint32_t module_id,
```

Comment:
My preference would be to leave "lib_code" support out until we have at least one module to test with. But if we do end up re-integrating this, can you please come up with better names - "code" doesn't tell me much at least

Reply:
does it make more send to use text, data, bss in the API naming ?

Reply:
@lgirdwood if I understood you correctly, this isn't about loading specific sections, it's about loading a complete module / library, but a library of a specific type - not one, implementing some module adapter driver for specific audio processing, but an auxiliary library to be used by user modules, like fft, math, etc.

Reply:
lyakh is right. I'll try to find a better name for these functions.